Friends bug fix

jason1794c opened the pull request:
The original methods used to retrieve the suggested users and friends list were PUT methods. A bug that appeared was that the current user's data was displayed as a friend card in the "Followed" tab on first render of the friends page. This pull request changes the fetching methods from PUT to GET.

jason1794c committed:
Render Navbar component in App.js rather than in each individual component

jason1794c committed:
Remove console log

jason1794c committed:
Delete UserContext.js

jason1794c committed:
Remove unused code

jason1794c committed:
Map the current user's list of friends and parse each of the JSON string of friends

Prior to this change, the list of friends only contained the followed user's email. Since the list of friends now contain JSON strings of followed users, we must parse each JSON string back into a proper JSON object. With the new parsedListOfFriends, the .some() function will look to see if the current user is following any users with the same email as the other user's profile.

jason1794c committed:
Update loginUser function to store the user's list of friends to the localStorage and removed the getFriends function

When the user logs in, the list of friends will be updated into the local storage. In addition, the getFriends function has been removed with its functionality moved to FollowingFriends.js since this function is only ever called when the user looks at their followed list.

jason1794c committed:
Move getFriends() function from action.js to here and changed fetch method from PUT to GET

The getFriends() function is moved from the actions.js file to here. In addition, the getFriends() function method was changed from using a PUT method to fetch data to using a GET method. Since the fetched data of followed users is an array of JSON strings, the friendDetails array was mapped twice. The first map is done to parse the JSON string back to a JSON object and the second map is to map the FriendCard component with the proper friend details.

jason1794c committed:
Adjusted code formatting

jason1794c committed:
Change method from PUT to GET

Prior to this change, the users list was fetched via a PUT method. This is changed to utilize the GET method in order to fetch the list of users.

jason1794c committed:
Remove EmailField from the friends attribute's ListField

jason1794c committed:
Update the friend variable in the FriendApi class

The friend variable in both the get and put methods have been updated to become User objects. In the PUT method, the friend object is appended in the current user's friends list.

jason1794c committed:
Remove space for formatting

jason1794c committed:
Remove unused code and console log for formatting

jason1794c committed:
Create a list the followed users' email and used it to filter the suggested users list

A list of emails based on the current user's followed list is created so that it can be used to filter the suggested users list. Any user that has their email in the current user's follow list will be filtered out of the suggested users list.

jason1794c committed:
Move Navbar component inside ProtectedRoute component so that the navbar won't show in the login and signup pages

jason1794c committed:
Remove console.log and formatted code spaces

jason1794c committed:
Add dependency to the useEffect hook so that the friends list will refresh when the user unfollows another user

shenziq1 reviewed (commented):
I've seen your changes, overall it's great but I have some disagreement.
1) You are not suppose to expose all users document to the frontend, i.e., for suggestedfriends, you are using the get method in usersAPI, which contains information for all users including passwords.
2) I found the friendAPI and userAPI have the same get method, you can delete the one in friendAPI and use the userAPI instead.
3) usersAPI's post and put methods are no longer needed and can be deleted.

MillanAir reviewed (commented):
I don't see my username when I navigate to friends page. Looks fine to me, however, Ziqi's review has some important points

jason1794c committed:
users variable retrieves a list of users without the users' password in the user object

Prior to this fix, the returned list of users contains all of the user information. The fix will return the same list of users, but without the password details.

jason1794c committed:
Switch fetching endpoint from /friends to /user

The GET request for the /friends endpoint is identical to the GET request for the /user endpoint. This switch was done because the reused code in the FriendsApi's GET method in the user_api.py file has been deleted.

jason1794c committed:
Delete UsersApi's PUT method and FriendApi's GET method

The deleted methods were not being used, and thus deleted.

jason1794c commented:
> I've seen your changes, overall it's great but I have some disagreement.
> 
> 1. You are not suppose to expose all users document to the frontend, i.e., for suggestedfriends, you are using the get method in usersAPI, which contains information for all users including passwords.
> 2. I found the friendAPI and userAPI have the same get method, you can delete the one in friendAPI and use the userAPI instead.
> 3. usersAPI's post and put methods are no longer needed and can be deleted.

The changes have been made.

jason1794c committed:
Merge branch 'dev' into friends-bug-fix
